derive_miniconf/src/lib.rs

```diff
@@ -174,14 +211,30 @@ fn derive_struct(mut typedef: TypeDefinition, data: syn::DataStruct, atomic: boo
         let field_name = &f.ident;
         quote! {
             #i => {
-                topic.push_str(concat!("/", stringify!(#field_name))).unwrap();
-                if let Some(()) = self.#field_name.recursive_iter(&mut index[1..], topic, value) {
+                let postfix = if topic.len() != 0 {
+                    concat!("/", stringify!(#field_name))
+                } else {
+                    stringify!(#field_name)
+                };
+
+                if topic.push_str(postfix).is_err() {
+                    return None;
+                }
+
+                if self.#field_name.recursive_iter(&mut index[1..], topic).is_some() {
                     return Some(());
                 }
-                else {
-                    index[0] += 1;
-                    index[1..].iter_mut().for_each(|x| *x = 0);
+
+                // Strip off the previously prepended index, since we completed that element and need
+                // to instead check the next one.
+                while let Some(character) = topic.pop() {
+                    if character == '/' {
+                        break;
+                    }
```

Comment:
We could backup the size of the topic before we push the new postfix, and then use `String::truncate()` to revert.

Reply:
I like that much more than the current approach - good idea, will implement